### PR TITLE
CBG-4411: Improvements to Diagnostic Sync Function dry run endpoint

### PR DESCRIPTION
**Pull request by RIT3shSapata**
CBG-4411

Describe your PR here...
- The endpoint can now run with a custom function provided by the user as well as a document
- The `SyncFnDryRun` function will now create an ephemeral `SyncRunner` if a custom sync function is passed
- Refactored the unit tests for the Sync Function dry run

## Pre-review checklist
- [x] Removed debug logging (`fmt.Print`, `log.Print`, ...)
- [x] Logging sensitive data? Make sure it's tagged (e.g. `base.UD(docID)`, `base.MD(dbName)`)
- [x] Updated relevant information in the API specifications (such as endpoint descriptions, schemas, ...) in `docs/api`

## Dependencies (if applicable)
- [ ] Link upstream PRs
- [ ] Update Go module dependencies when merged

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGatewayIntegration/build?delay=0sec)
- [x] `GSI=true,xattrs=true` https://jenkins.sgwdev.com/job/SyncGatewayIntegration/202/
- [ ] `GSI=true,xattrs=true` https://jenkins.sgwdev.com/job/SyncGatewayIntegration/207/

### CODE REVIEW

File: db/crud.go [hunk 1]
```diff
@@ -1695,7 +1695,7 @@ func (db *DatabaseCollectionWithUser) PutExistingRevWithBody(ctx context.Context
 }
 
 // SyncFnDryrun Runs a document through the sync function and returns expiry, channels doc was placed in, access map for users, roles, handler errors and sync fn exceptions
-func (db *DatabaseCollectionWithUser) SyncFnDryrun(ctx context.Context, body Body, docID string) (*channels.ChannelMapperOutput, error, error) {
+func (db *DatabaseCollectionWithUser) SyncFnDryrun(ctx context.Context, body Body, docID, syncFn string) (*channels.ChannelMapperOutput, error, error) {
```
Comment (torcolvin) [inline db/crud.go]: Let's document the behavior of this function and add a pre condition to help the callers do the right thing:

```suggestion
// SyncFnDryrun Runs a document through the sync function and returns expiry, channels doc was placed in, access map for users, roles, handler errors and sync fn exceptions.
// If a docID is a non empty string, the document will be fetched from the bucket, otherwise the body will be used. If both are specified, this function returns an error.
// The first error return value represents an error that occurs before the sync function is run. The second error return value represents an exception from the sync function.
func (db *DatabaseCollectionWithUser) SyncFnDryrun(ctx context.Context, body Body, docID, syncFn string) (*channels.ChannelMapperOutput, error, error) {
	if docID != "" && len(body) > 0 {
		return nil, fmt.Errorf("cannot call SyncFnDryrun with both a body and a docID"), nil
```

A better abstraction for this is to read the docID before you call this function and then pass the body into this function, so you don't have multiple pathways.

It is also extremely weird to return two errors, and that could be changed but is a more complicated change. I think it would be good to write a test that would return different types of errors on this function and see what they are. Two ways of changing this could be:

1. Create a different struct that like:

```
type SyncDryRunOutput {
	output *channels.ChannelMapperOutput
    err error // represents an exception running the sync function
}
```

2. Wrap the error in a specific error type like:

`const ErrSyncFn = errors.New("Exception occurred while calling a sync function")`

Then if `Call` returns an exception, return `fmt.Errorf("%w: %w", err)` and you can use `errors.Is` by the caller.

I think changing the error might be out of scope for this change, but I think it's important to make sure there's a test for that code pathway.

File: db/crud.go [hunk 2]
```diff
@@ -1704,7 +1704,7 @@ func (db *DatabaseCollectionWithUser) SyncFnDryrun(ctx context.Context, body Bod
 	if docID != "" {
 		if docInBucket, err := db.GetDocument(ctx, docID, DocUnmarshalAll); err == nil {
 			oldDoc = docInBucket
-			if doc._body == nil {
+			if doc._body == nil || len(doc._body) == 0 {
```
Comment (torcolvin) [inline db/crud.go]: nit about go: checking for nil is redundant

```suggestion
			if len(doc._body) == 0 {
```

File: rest/diagnostic_doc_api.go [hunk 3]
```diff
@@ -69,15 +75,19 @@ func (h *handler) handleGetDocChannels() error {
...
 	docid := h.getQuery("doc_id")
+	contentType, _, _ := mime.ParseMediaType(h.rq.Header.Get("Content-Type"))
 
-	body, err := h.readDocument()
-	if err != nil {
-		if docid == "" {
-			return fmt.Errorf("no doc id provided for dry run and error reading body: %s", err)
-		}
+	if contentType != "application/json" && contentType != "" {
+		return base.HTTPErrorf(http.StatusUnsupportedMediaType, "Invalid Content type %s", contentType)
```
Comment (torcolvin) [inline rest/diagnostic_doc_api.go]:
```suggestion
		return base.HTTPErrorf(http.StatusUnsupportedMediaType, "Invalid Content-Type header: %s. Needs to be empty or application/json", contentType)
```

File: rest/diagnostic_doc_api_test.go
Comment (torcolvin) [inline rest/diagnostic_doc_api_test.go:71]:
```suggestion
	rt := NewRestTester(t, &RestTesterConfig{
		PersistentConfig: true,
		ImportFilter: function(doc) { if (doc.user.num) { return true; } else { return false; } }"},
		SyncFn: "function(doc,oldDoc){if (doc.user.num >= 100) {channel(doc.channel);} else {throw({forbidden: 'user num too low'});}if (oldDoc){ console.log(oldDoc); if (oldDoc.user.num > doc.user.num) { access(oldDoc.user.name, doc.channel);} else {access(doc.user.name[0], doc.channel);}}}"`
		})
```

Using this format will allow you to call `NewDbConfig` below and have this work with collections.

File: rest/diagnostic_doc_api_test.go [hunk 4]
```diff
@@ -72,80 +72,18 @@ func TestGetDocDryRuns(t *testing.T) {
...
-
-	err = json.Unmarshal(response.BodyBytes(), &respMap)
-	assert.NoError(t, err)
-	assert.Equal(t, respMap.Roles, channels.AccessMap{"user": channels.BaseSetOf(t, "role1")})
 	newSyncFn := `"function(doc,oldDoc){if (doc.user.num >= 100) {channel(doc.channel);} else {throw({forbidden: 'user num too low'});}if (oldDoc){ console.log(oldDoc); if (oldDoc.user.num > doc.user.num) { access(oldDoc.user.name, doc.channel);} else {access(doc.user.name[0], doc.channel);}}}"`
-	resp = rt.SendAdminRequest("POST", "/db/_config", fmt.Sprintf(
+	resp := rt.SendAdminRequest("PUT", "/db/", fmt.Sprintf(
 		`{"bucket":"%s", "num_index_replicas": 0, "enable_shared_bucket_access": %t, "sync":%s, "import_filter":%s}`,
 		bucket, base.TestUseXattrs(), newSyncFn, ImportFilter))
 	RequireStatus(t, resp, http.StatusCreated)
```
Comment (torcolvin) [inline rest/diagnostic_doc_api_test.go:79]: To work with collections see above for `RestTesterConfig` change.

```suggestion
resp := rt.CreateDatabase("db", rt.NewDbConfig())`
```

**Review by torcolvin — COMMENTED** [review]
I'm going to review test cases later today, I wanted to give some comments.

File: db/crud.go (hunk 2 above)
Comment (RIT3shSapata) [reply db/crud.go]: done

File: db/crud.go (hunk 1 above)
Comment (RIT3shSapata) [reply db/crud.go]: I've made the changes the changes to the documentation. I do agree that it is strange that there are two errors are being returned, please let me know if you would like me to make the changes to the way error is handled. 

The first method of resolving this, already exists. In the handler there is a struct `SyncFnDryRun` which stores the error returned from the Sync Function and stores it in the `Exception` property. 

There are already test cases in the unit tests which try to cover the codepath of when there is an error in sync function. Please check test cases 6,7 and 11. Did you want coverage similar to the ones in these cases or a different one?

Comment (torcolvin) [reply db/crud.go]: I think we can leave this weird code pathway for now in favor of getting things done for this sprint, but if it ends up causing confusion while you are working I think it might make sense to fix?

The answer sort of depends on how long you think it would take you to fix, and I think it is okay to leave it.

Comment (RIT3shSapata) [reply db/crud.go]: Fixed it in the latest commit.

**Comment by RIT3shSapata** [comment]
Hey @bbrks @torcolvin , I've refactored bit of redundant code, which was pointed out @torcolvin . I've also updated the API docs for the diagnostic API as well.

File: db/crud.go [hunk 5]
```diff
@@ -1694,37 +1694,18 @@ func (db *DatabaseCollectionWithUser) PutExistingRevWithBody(ctx context.Context
...
-		} else {
-			return nil, err, nil
-		}
-	} else {
-		oldDoc._body = nil
-	}
+// SyncFnDryrun Runs a document through the sync function and returns expiry, channels doc was placed in, access map for users, roles, handler errors and sync fn exceptions.
+// If a docID is a non empty string, the document will be fetched from the bucket, otherwise the body will be used. If both are specified, this function returns an error.
+// The first error return value represents an error that occurs before the sync function is run. The second error return value represents an exception from the sync function.
+func (db *DatabaseCollectionWithUser) SyncFnDryrun(ctx context.Context, oldDoc *Document, body Body, docID, syncFn string) (*channels.ChannelMapperOutput, error) {
```
Comment (bbrks) [inline db/crud.go]: Similarly, `docID` is now unused and is making the usage of the function ambiguous without reading the implementation.

```suggestion
func (db *DatabaseCollectionWithUser) SyncFnDryrun(ctx context.Context, oldDoc *Document, body Body, syncFn string) (*channels.ChannelMapperOutput, error) {
```

File: db/crud.go [hunk 6]
```diff
@@ -1694,37 +1694,18 @@ func (db *DatabaseCollectionWithUser) PutExistingRevWithBody(ctx context.Context
...
-			doc._body[BodyRev] = oldDoc.SyncData.GetRevTreeID()
-		} else {
-			return nil, err, nil
-		}
-	} else {
-		oldDoc._body = nil
-	}
+// SyncFnDryrun Runs a document through the sync function and returns expiry, channels doc was placed in, access map for users, roles, handler errors and sync fn exceptions.
+// If a docID is a non empty string, the document will be fetched from the bucket, otherwise the body will be used. If both are specified, this function returns an error.
+// The first error return value represents an error that occurs before the sync function is run. The second error return value represents an exception from the sync function.
```
Comment (bbrks) [inline db/crud.go]: This comment is outdated since it's now up to the caller to pass in document body.

Also only returns one error.

```suggestion
// SyncFnDryrun Runs the given document body through a sync function and returns expiry, channels doc was placed in, access map for users, roles, handler errors and sync fn exceptions.
// If syncFn is provided, it will be used instead of the one configured on the database.
```

File: db/crud.go [hunk 7]
```diff
@@ -1768,16 +1749,34 @@ func (db *DatabaseCollectionWithUser) SyncFnDryrun(ctx context.Context, body Bod
...
 	if err != nil {
-		return nil, err, nil
+		return nil, err
+	}
+	var output *channels.ChannelMapperOutput
+	var syncErr error
+	if syncFn == "" {
+		output, err = db.ChannelMapper.MapToChannelsAndAccess(ctx, mutableBody, string(oldDoc._rawBody), metaMap, syncOptions)
+		if err != nil {
+			syncErr = fmt.Errorf("%s%s", base.ErrSyncFnDryRun, err)
```
Comment (bbrks) [inline db/crud.go]: ~I'd be tempted to use `%w` instead of `%s` here for `err` so the original can still be used by `errors.Is`~

```suggestion
			syncErr = fmt.Errorf("%s %w", base.ErrSyncFnDryRun, err)
```

**Edit:** This comment is superseded by the error handling comment in the REST API handler function.

File: db/crud.go [hunk 8]
```diff
@@ -1768,16 +1749,34 @@ func (db *DatabaseCollectionWithUser) SyncFnDryrun(ctx context.Context, body Bod
...
+		}
+	} else {
+		jsTimeout := time.Duration(base.DefaultJavascriptTimeoutSecs) * time.Second
+		syncRunner, err := channels.NewSyncRunner(ctx, syncFn, jsTimeout)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create sync runner: %v", err)
```
Comment (bbrks) [inline db/crud.go:1723]:
```suggestion
			return nil, fmt.Errorf("failed to create sync runner: %w", err)
```

**Edit:** This comment is superseded by the error handling comment in the REST API handler function.

File: db/crud.go [hunk 9]
```diff
@@ -1775,7 +1775,22 @@ func (db *DatabaseCollectionWithUser) SyncFnDryrun(ctx context.Context, body Bod
 	if err != nil {
 		return nil, err, nil
 	}
-	output, err := db.ChannelMapper.MapToChannelsAndAccess(ctx, mutableBody, string(oldDoc._rawBody), metaMap, syncOptions)
+	base.InfofCtx(ctx, base.KeyHTTP, "old doc: %s, new doc: %s", string(oldDoc._rawBody), mutableBody)
+	var output *channels.ChannelMapperOutput
+	if syncFn == "" {
+		output, err = db.ChannelMapper.MapToChannelsAndAccess(ctx, mutableBody, string(oldDoc._rawBody), metaMap, syncOptions)
+	} else {
+		jsTimeout := time.Duration(base.DefaultJavascriptTimeoutSecs) * time.Second
```
[the remark this answers is not in the document]
Comment (bbrks) [reply db/crud.go:1720]: It's fine as-is

File: rest/diagnostic_doc_api.go [hunk 10]
```diff
@@ -69,24 +76,47 @@ func (h *handler) handleGetDocChannels() error {
 // If docid is specified and the document does not exist in the bucket, it will return error
 func (h *handler) handleSyncFnDryRun() error {
 	docid := h.getQuery("doc_id")
+	contentType, _, _ := mime.ParseMediaType(h.rq.Header.Get("Content-Type"))
```
Comment (bbrks) [inline rest/diagnostic_doc_api.go]: `h.readJSONInto` already does this mime type checking (via the `processContentEncoding` function) and handles things like gzip compression.

```suggestion
```

File: docs/api/paths/diagnostic/keyspace-sync.yaml [hunk 11]
```diff
@@ -7,16 +7,39 @@
 # the file licenses/APL2.txt.
 parameters:
   - $ref: ../../components/parameters.yaml#/keyspace
```
Comment (bbrks) [inline docs/api/paths/diagnostic/keyspace-sync.yaml:9]: We should include the available query parameters in here (`doc_id`) - appreciate this was already missing before this PR but let's fix it up now.

```suggestion
  - $ref: ../../components/parameters.yaml#/keyspace
  - $ref: ../../components/parameters.yaml#/doc_id
```

File: rest/diagnostic_doc_api.go [hunk 12]
```diff
@@ -69,15 +75,19 @@ func (h *handler) handleGetDocChannels() error {
...
-	if err != nil {
-		if docid == "" {
-			return fmt.Errorf("no doc id provided for dry run and error reading body: %s", err)
-		}
+	if contentType != "application/json" && contentType != "" {
+		return base.HTTPErrorf(http.StatusUnsupportedMediaType, "Invalid Content type %s", contentType)
+	}
+
+	var syncDryRunPayload SyncFnDryRunPayload
+	err := h.readJSONInto(&syncDryRunPayload)
```
[the remark this answers is not in the document]
Comment (bbrks) [reply rest/diagnostic_doc_api.go:80]: I agree with this comment.

Based on the table in the API documentation describing how this endpoint works - in the case where you provide a doc ID to act as `oldDoc`, and provide some invalid JSON for `doc` - this error will be a silent failure (running the sync function on an empty body).

This should be able to be covered by a unit test case.

File: rest/diagnostic_doc_api.go [hunk 13]
```diff
@@ -69,24 +76,47 @@ func (h *handler) handleGetDocChannels() error {
...
-		return err
-	}
-	if syncFnErr != nil {
-		resp := SyncFnDryRun{
-			Exception: syncFnErr.Error(),
+		if strings.Contains(err.Error(), base.ErrSyncFnDryRun.Error()) {
+			errMsg := strings.ReplaceAll(err.Error(), base.ErrSyncFnDryRun.Error(), "")
+			resp := SyncFnDryRun{
+				Exception: errMsg,
+			}
```
Comment (bbrks) [inline rest/diagnostic_doc_api.go]: Something like this would be better achieved using `errors.As`, where you assert on the type of error being returned instead of string comparisons.

https://go.dev/blog/go1.13-errors

This would involve making `ErrSyncFnDryRun` a new type instead of an instance of `sgError`:

```go
const syncFnDryRunErrorPrefix = "Error returned from Sync Function"

// SyncFnDryRunError is returned when the sync function dry run returns an error.
// It wraps the original error for errors.Is and the type supports errors.As
type SyncFnDryRunError struct {
	Err error
}

func (e *SyncFnDryRunError) Error() string {
	if e == nil {
		return syncFnDryRunErrorPrefix
	}
	if e.Err == nil {
		return syncFnDryRunErrorPrefix
	}
	return syncFnDryRunErrorPrefix + ": " + e.Err.Error()
}

func (e *SyncFnDryRunError) Unwrap() error {
	if e == nil {
		return nil
	}
	return e.Err
}
```

In use:
```go
return &base.SyncFnDryRunError{Err: err}
```

This then allows the error handling to use `errors.As` and still grab the underlying error value.

```go
	output, err := h.collection.SyncFnDryrun(h.ctx(), oldDoc, syncDryRunPayload.Doc, docid, syncDryRunPayload.Function)
	if err != nil {
		var syncFnDryRunErr *base.SyncFnDryRunError
		if !errors.As(err, &syncFnDryRunErr) {
			return err
		}

		errMsg := syncFnDryRunErr.Error()
		if syncFnDryRunErr.Unwrap() != nil {
			errMsg = syncFnDryRunErr.Err.Error()
		}
		resp := SyncFnDryRun{
			Exception: errMsg,
		}
		h.writeJSON(resp)
		return nil
	}
```

It's a little more code, but less brittle than string comparisons and is more idiomatic (allows the error/type to play nicely with the stdlib errors package)

File: rest/diagnostic_doc_api_test.go [hunk 14]
```diff
@@ -1079,3 +998,241 @@ func TestGetUserDocAccessDuplicates(t *testing.T) {
...
+		document        string
+		docID           string
+		existingDoc     bool
+		existingDocID   string
+		existingDocBody string
+		expectedOutput  SyncFnDryRun
+		expectedStatus  int
+	}{
+		{
+			name:           "test_case_1",
```
Comment (bbrks) [inline rest/diagnostic_doc_api_test.go]: It would be useful to have these test names reflect what is being tested for each subtest (e.g. For this one, since the provided sync function is identical to the database one, we can't determine which ran. What are we testing here?)

This helps us ensure we've got coverage of edge cases and makes it easier to figure out what is going on when these tests need to be looked at.

**Review by bbrks — CHANGES_REQUESTED** [review]
Let me know if you want to pair up or talk through any of these comments - I'd be happy to

File: db/crud.go (hunk 5 above)
Comment (RIT3shSapata) [reply db/crud.go]: doc ID is required for the preprocessing of the newDoc. I've moved all the preprocessing inside the handler. I wanted to know if its okay to do all the preprocessing in the handler instead of the Dry run method.

File: db/crud.go [hunk 15]
```diff
@@ -1694,90 +1694,43 @@ func (db *DatabaseCollectionWithUser) PutExistingRevWithBody(ctx context.Context
...
+		syncRunner, err := channels.NewSyncRunner(ctx, syncFn, jsTimeout)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create sync runner: %v", err)
+		}
+		jsOutput, err := syncRunner.Call(ctx, mutableBody, string(oldDoc._rawBody), metaMap, syncOptions)
+		if err != nil {
+
+			return nil, fmt.Errorf("failed to create sync runner: %v", err)
```
Comment (bbrks) [inline db/crud.go]: Does this need to be a typed `SyncFnDryRunError` as well? The error is a runtime error from the javascript execution not creating a sync runner.

It seems like we may end up with slightly different behaviour between a custom sync function and the actual sync function because of this.

File: rest/diagnostic_doc_api.go [hunk 16]
```diff
@@ -70,20 +76,89 @@ func (h *handler) handleGetDocChannels() error {
 func (h *handler) handleSyncFnDryRun() error {
 	docid := h.getQuery("doc_id")
 
-	body, err := h.readDocument()
+	var syncDryRunPayload SyncFnDryRunPayload
+	err := h.readJSONInto(&syncDryRunPayload)
+	// Only require a valid JSON payload if docid is not provided.
+	// If docid is provided, the sync function will use the document from the bucket, and the payload is optional.
```
Comment (bbrks) [inline rest/diagnostic_doc_api.go]: These comments are from the old behaviour now and can be removed.

**Review by bbrks — CHANGES_REQUESTED** [review]
This looks mostly good until the CBG-5061 refactor - mainly just one comment about the returned error type for a custom sync function and then we can get this merged.

File: rest/diagnostic_doc_api.go [hunk 17]
```diff
@@ -70,20 +76,89 @@ func (h *handler) handleGetDocChannels() error {
...
+	if err != nil {
+		var syncFnDryRunErr *base.SyncFnDryRunError
+		if !errors.As(err, &syncFnDryRunErr) {
+			return err
+		}
+
+		errMsg := syncFnDryRunErr.Error()
+		if syncFnDryRunErr.Unwrap() != nil {
+			errMsg = syncFnDryRunErr.Err.Error()
+		}
```
Comment (bbrks) [inline rest/diagnostic_doc_api.go]: Stole comment from the import PR since I spotted it there first.

---

I don't think the `Unwrap` step isn't necessary since `Error()` on the typed error returns the underlying error (if present) - and if we have a nil typed error, we can use that without a problem since it returns `syncFnDryRunErrorPrefix` as a generic error message.


```suggestion
		errMsg := syncFnDryRunErr.Error()
```

**Review by bbrks — APPROVED** [review]
LGTM 👍 thanks!
